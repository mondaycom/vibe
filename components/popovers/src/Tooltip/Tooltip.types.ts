export type TooltipPositions = "top" | "right" | "bottom" | "left";

export type TooltipTheme = "dark" | "primary";
