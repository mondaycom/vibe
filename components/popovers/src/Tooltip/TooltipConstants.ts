/**
 * @deprecated
 */
export enum TooltipPositions {
  TOP = "top",
  RIGHT = "right",
  BOTTOM = "bottom",
  LEFT = "left"
}

/**
 * @deprecated
 */
export enum TooltipTheme {
  Dark = "dark",
  Primary = "primary"
}
