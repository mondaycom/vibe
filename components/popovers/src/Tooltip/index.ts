export { default as Tooltip, type TooltipProps } from "./Tooltip";

export * from "./Tooltip.types";
