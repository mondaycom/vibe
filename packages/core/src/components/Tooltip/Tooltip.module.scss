@import "../../styles/typography";

.tooltip {
  position: relative;
  display: inline-block;
  border-radius: var(--tooltip-border-radius, var(--border-radius-small));
  box-shadow: var(--box-shadow-medium);
  @include font-input();
  max-width: var(--tooltip-max-width, 50vw);
  word-break: break-word;
  background-color: var(--inverted-color-background);
  color: var(--text-color-on-inverted);
  --tooltip-max-width: 240px;
  white-space: pre-wrap;

  .image {
    display: block;
    position: relative;
    border-top-right-radius: 2px;
    border-top-left-radius: 2px;
    padding: 2px 2px 0 2px;
    object-fit: cover;
    width: 100%;
    height: 100%;
    min-width: var(--tooltip-max-width);
    z-index: 1;
  }

  .title {
    @include vibe-text(text2, bold);
  }

  .content {
    word-break: break-word;
    display: inline-block;
    padding: var(--tooltip-padding, #{var(--spacing-small) var(--spacing-medium)});
  }
}

.tooltip.paddingSizeMd {
  border-radius: var(--border-radius-medium);
  padding: var(--spacing-medium);
  @include font-input();
}

.tooltip a.tooltipWhiteLink {
  color: #fff;
}

.arrow,
.dark,
.arrow.dark {
  background-color: var(--inverted-color-background);
  color: var(--text-color-on-inverted);
}

.arrow.arrow.dark,
.arrow.arrow.primary {
  :global(.dark-app-theme) &,
  :global(.black-app-theme) &,
  :global(.hacker-app-theme) & {
    box-shadow: none;
  }
}

.primary,
.arrow.primary {
  background-color: var(--primary-color);
  color: var(--text-color-on-primary);
}
