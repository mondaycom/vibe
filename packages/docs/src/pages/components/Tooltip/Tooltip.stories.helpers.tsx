import React from "react";
import { StorybookLink, Tip } from "vibe-storybook-components";

export const modifiers = [
  {
    name: "preventOverflow",
    options: {
      mainAxis: false
    }
  },
  {
    name: "flip",
    options: {
      // @ts-ignore
      fallbackPlacements: []
    }
  }
];

export const TipOtherComponents = () => (
  <Tip>
    As tooltips only surface from a hover, never include links or buttons in the copy. If your tooltip requires either
    of these, considers putting your content in a{" "}
    <StorybookLink page="Components/AttentionBox" size={StorybookLink.sizes.SMALL}>
      Attention box
    </StorybookLink>{" "}
    or{" "}
    <StorybookLink page="Components/Dialog" size={StorybookLink.sizes.SMALL}>
      Dialog.
    </StorybookLink>
  </Tip>
);
