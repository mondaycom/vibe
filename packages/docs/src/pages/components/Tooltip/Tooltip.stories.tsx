import React from "react";
import { Tooltip, type TooltipProps, Button, Flex, IconButton } from "@vibe/core";
import { Hide, Menu, Subitems } from "@vibe/icons";
import { modifiers } from "./Tooltip.stories.helpers";
import image from "./assets/tooltip-image.png";
import { createStoryMetaSettingsDecorator } from "../../../utils/createStoryMetaSettingsDecorator";

const metaSettings = createStoryMetaSettingsDecorator({
  component: Tooltip,
  iconPropNamesArray: ["icon"]
});

export default {
  title: "Components/Tooltip",
  component: Tooltip,
  argTypes: metaSettings.argTypes,
  decorators: metaSettings.decorators
};

export const Overview = {
  render: (args: TooltipProps) => (
    <div style={{ padding: "40px 0 0 50px" }}>
      <Tooltip
        id="overview-tooltip"
        {...args}
        open
        hideWhenReferenceHidden
        // The modifier's purpose is to prevent the tooltip from being displayed when the user scrolls the story upwards / downwards.
        // Therefore, there is no need to move this prop in your implementations.
        modifiers={[
          {
            name: "preventOverflow",
            options: {
              mainAxis: false
            }
          },
          {
            name: "flip",
            options: {
              // @ts-ignore
              fallbackPlacements: []
            }
          }
        ]}
      >
        <div id="overview-tooltip-trigger" />
      </Tooltip>
    </div>
  ),
  name: "Overview",
  args: {
    shouldShowOnMount: true,
    content: "I'm a tooltip"
  },
  parameters: {
    chromatic: {
      pauseAnimationAtEnd: true
    }
  }
};

export const TooltipWithTitle = () => (
  <div style={{ padding: "30px 0" }}>
    <Tooltip
      id="tooltip-with-title"
      content="Hidden columns"
      title="Tooltip title"
      shouldShowOnMount
      position="right"
      hideWhenReferenceHidden
      // The modifier's purpose is to prevent the tooltip from being displayed when the user scrolls the story upwards / downwards.
      // Therefore, there is no need to move this prop in your implementations.
      modifiers={[
        {
          name: "preventOverflow",
          options: {
            mainAxis: false
          }
        },
        {
          name: "flip",
          options: {
            // @ts-ignore
            fallbackPlacements: []
          }
        }
      ]}
      open
    >
      <div id="tooltip-with-title-trigger" />
    </Tooltip>
  </div>
);

export const TooltipWithImage = () => (
  <div style={{ padding: "120px 0" }}>
    <Tooltip
      id="tooltip-with-image"
      content="Max width tooltip with a long text example"
      title="Tooltip title"
      shouldShowOnMount
      image={image}
      position="right"
      hideWhenReferenceHidden
      modifiers={modifiers}
      style={{ minHeight: "135px" }}
      open
    >
      <div id="tooltip-with-image-trigger" />
    </Tooltip>
  </div>
);

export const Positions = {
  render: () => {
    // The modifier's purpose is to prevent the tooltip from being displayed when the user scrolls the story upwards / downwards.
    // Therefore, there is no need to move this prop in your implementations.
    const modifiers = [
      {
        name: "preventOverflow",
        options: {
          mainAxis: false
        }
      },
      {
        name: "flip",
        options: {
          // @ts-ignore
          fallbackPlacements: []
        }
      }
    ];

    return (
      <Flex gap={16}>
        <div style={{ padding: "0 64px 68px 0", margin: "0 32px" }}>
          {/* The modifier's purpose is to prevent the tipseen from being displayed when the user scrolls the story upwards / downwards.
         Therefore, there is no need to move this prop in your implementations. */}
          <Tooltip
            id="position-top-tooltip"
            hideWhenReferenceHidden
            modifiers={modifiers}
            content="Top"
            shouldShowOnMount
            position="bottom"
            open
          >
            <div />
          </Tooltip>
        </div>
        <div style={{ padding: "50px 0 0 0", margin: "0 32px" }}>
          <Tooltip
            id="position-bottom-tooltip"
            hideWhenReferenceHidden
            modifiers={modifiers}
            content="Bottom"
            shouldShowOnMount
            open
          >
            <div />
          </Tooltip>
        </div>
        <div style={{ padding: "0 32px 8px 32px", margin: "0 32px" }}>
          <Tooltip
            id="position-left-tooltip"
            hideWhenReferenceHidden
            modifiers={modifiers}
            content="Left"
            position="right"
            shouldShowOnMount
            open
          >
            <div />
          </Tooltip>
        </div>
        <div style={{ padding: "0 64px 8px 64px", margin: "0 64px" }}>
          <Tooltip
            id="position-right-tooltip"
            hideWhenReferenceHidden
            modifiers={modifiers}
            content="Right"
            position="left"
            shouldShowOnMount
            open
          >
            <div />
          </Tooltip>
        </div>
      </Flex>
    );
  },
  name: "Positions",
  parameters: {
    chromatic: {
      pauseAnimationAtEnd: true
    }
  }
};

export const IconTooltip = () => (
  <Tooltip id="icon-tooltip" hideWhenReferenceHidden content="Hidden columns">
    <IconButton id="icon-tooltip-trigger" ariaLabel="Hide columns" kind="secondary" size="small" icon={Hide} />
  </Tooltip>
);

export const DefinitionTooltip = () => (
  <Tooltip id="definition-tooltip" hideWhenReferenceHidden content="Item name: Bottom sheets">
    <Button id="definition-tooltip-trigger" ariaLabel="View subitems" kind="secondary" size="small" leftIcon={Subitems}>
      Subitem
    </Button>
  </Tooltip>
);

export const ImmediateTooltips = () => (
  <Flex gap="small">
    <Tooltip id="immediate-tooltip-1" hideWhenReferenceHidden immediateShowDelay={0} content="I'm a tooltip">
      <IconButton id="immediate-trigger-1" ariaLabel="Menu button 1" icon={Menu} kind="secondary" size="small" />
    </Tooltip>
    <Tooltip id="immediate-tooltip-2" hideWhenReferenceHidden immediateShowDelay={0} content="I'm a tooltip">
      <IconButton id="immediate-trigger-2" ariaLabel="Menu button 2" icon={Menu} kind="secondary" size="small" />
    </Tooltip>
    <Tooltip id="immediate-tooltip-3" hideWhenReferenceHidden content="I'm a tooltip">
      <IconButton id="immediate-trigger-3" ariaLabel="Menu button 3" icon={Menu} kind="secondary" size="small" />
    </Tooltip>
  </Flex>
);
