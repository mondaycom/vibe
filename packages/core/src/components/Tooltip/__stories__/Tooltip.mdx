import Tooltip from "../Tooltip";
import { Canvas, Meta } from "@storybook/blocks";
import { Bolt } from "@vibe/icons";
import Icon from "../../Icon/Icon";
import Button from "../../Button/Button";
import TabList from "../../Tabs/TabList/TabList";
import Tab from "../../Tabs/Tab/Tab";
import Link from "../../Link/Link.tsx";
import IconButton from "../../IconButton/IconButton";
import { CHIP, LABEL, TIPSEEN } from "../../../storybook/components/related-components/component-description-map";
import * as TooltipStories from "./Tooltip.stories";
import { TipOtherComponents, modifiers } from "./Tooltip.stories.helpers";
import image from "./assets/tooltip-image.png";

<Meta of={TooltipStories} />

# Tooltip

Tooltips provide helpful information and can include detailed context about a specific component when a user hovers on it.

<Canvas of={TooltipStories.Overview} />

### Import

```js
import { Tooltip } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Should never contain critical information that a user would definitely need to proceed.",
    "Use when the description content would be too much information to include inline or create too much clutter. For example, when expert users have seen it many times.",
    "Only show 1 tooltip at a time.",
    "Whenever text is shortened and an ellipsis (...) appears, add a tooltip that contains the full version of the text."
  ]}
/>

<TipOtherComponents />

## Variants

### Tooltip with title

Use tooltip with title when you want to present more complex content and you need to add a heading to emphasize the main idea. Titles can also come with an icon.

<Canvas of={TooltipStories.TooltipWithTitle} />

### Tooltip with image

Use tooltip with image to provide a preview for an image or when giving a closer look at a feature.

<Canvas of={TooltipStories.TooltipWithImage} />

### Positions

Tooltip's arrow can appear from top, bottom, left or right.

<Canvas of={TooltipStories.Positions} />

## Do's and Don'ts

<ComponentRules
  rules={[
    {
      positive: {
        component: (
          <Tooltip shouldShowOnMount content="Quick search + B" modifiers={modifiers}>
            <IconButton kind="secondary" icon={Bolt} iconSize="32" />
          </Tooltip>
        ),
        description: "Use tooltips for interactive imagery."
      },
      negative: {
        component: (
          <Tooltip shouldShowOnMount content="Quick search + B" modifiers={modifiers}>
            <Button>Quick search</Button>
          </Tooltip>
        ),
        description: "Don't use tooltips to restate visible UI text."
      }
    },
    {
      positive: {
        component: (
          <TabList>
            <Tab>Boards</Tab>
            <Tab>Items</Tab>
          </TabList>
        ),
        description: "If an element doesn't need additional clarification, don't use a tooltip."
      },
      negative: {
        component: (
          <TabList>
            <Tooltip shouldShowOnMount content="Boards" position="bottom" modifiers={modifiers}>
              <Tab>Boards</Tab>
            </Tooltip>
            <Tab>Items</Tab>
          </TabList>
        ),
        description: "Don't use tooltips to restate text that is already displayed within the element."
      }
    },
    {
      positive: {
        component: (
          <Tooltip
            shouldShowOnMount
            title="I'm a rich tooltip"
            modifiers={modifiers}
            open
            content={
              <div>
                I have a title, text, image and I can come <Link text="with a link" inlineText color="onPrimary" />.
              </div>
            }
          >
            <div style={{ position: "relative", top: "25%" }} />
          </Tooltip>
        ),
        description: "Add links inline and in the same color of the text."
      },
      negative: {
        component: (
          <Tooltip
            shouldShowOnMount
            title="I'm a rich tooltip"
            modifiers={modifiers}
            open
            content={
              <div>
                I have a title, text, image and I can come with a link. <Link text="Read more" />
              </div>
            }
          >
            <div style={{ position: "relative", top: "25%" }} />
          </Tooltip>
        ),
        description: "Don't add links on a separate line or in a different color from the text."
      }
    },
    {
      positive: {
        component: (
          <Tooltip
            content="Empower users to visualize data."
            shouldShowOnMount
            image={image}
            position="right"
            modifiers={modifiers}
            open
          >
            <div style={{ position: "relative", left: "-25%" }} />
          </Tooltip>
        ),
        description: "Use this app to visualize data."
      },
      negative: {
        component: (
          <Tooltip
            content="View our automation center."
            shouldShowOnMount
            image={image}
            position="right"
            modifiers={modifiers}
            open
          >
            <div style={{ position: "relative", left: "-25%" }} />
          </Tooltip>
        ),
        description: "Don't add an image just for aesthetic reasons, or if it doesn't help the user."
      }
    }
  ]}
/>

## Use cases and examples

### Icon tooltip

An icon tooltip is used to clarify the action or name of an interactive icon button. Provide tooltips for all unlabelled icons.

<Canvas of={TooltipStories.IconTooltip} />

### Definition tooltip

The definition tooltip provides additional help or defines an item or term. It may be used on the label of a UI element, or on a word embedded in a paragraph.

<Canvas of={TooltipStories.DefinitionTooltip} />

### Immediate tooltips

Immediately show when another is shown. The two left tooltips uses the immediate show prop, the right one ignores it and should always have show delay.

<Canvas of={TooltipStories.ImmediateTooltips} />

## Related components

<RelatedComponents componentsNames={[CHIP, LABEL, TIPSEEN]} />
