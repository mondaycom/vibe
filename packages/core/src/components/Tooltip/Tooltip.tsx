/* eslint-disable react/jsx-props-no-spreading */
import { camelCase, isFunction } from "lodash-es";
import cx from "classnames";
import React, { CSSProperties, isValidElement, PureComponent, ReactElement } from "react";
import { Modifier } from "react-popper";
import Dialog from "../Dialog/Dialog";
import { DialogAnimationType, DialogTriggerEvent } from "../Dialog/Dialog.types";
import { HideShowEvent as HideShowEventEnum, AnimationType as AnimationTypeEnum } from "../Dialog/DialogConstants";
import VibeComponentProps from "../../types/VibeComponentProps";
import { TooltipTheme as TooltipThemeEnum, TooltipPositions as TooltipPositionsEnum } from "./TooltipConstants";
import { ElementContent } from "../../types/ElementContent";
import { MoveBy } from "../../types/MoveBy";
import { getStyle } from "../../helpers/typesciptCssModulesHelper";
import styles from "./Tooltip.module.scss";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import { SubIcon } from "../../types";
import Icon from "../Icon/Icon";
import Flex from "../Flex/Flex";
import { TooltipPositions, TooltipTheme } from "./Tooltip.types";

export type TooltipProps = TooltipBaseProps & (TooltipWithChildrenProps | TooltipWithoutChildrenProps);

interface TooltipWithoutChildrenProps {
  /**
   * If true, the tooltip will be rendered even if there are no children.
   */
  forceRenderWithoutChildren: boolean;
  /**
   * The children elements that the tooltip is attached to.
   */
  children?: ReactElement | Array<ReactElement>;
}

interface TooltipWithChildrenProps {
  /**
   * If true, the tooltip will be rendered even if there are no children.
   */
  forceRenderWithoutChildren?: boolean;
  /**
   * The children elements that the tooltip is attached to.
   */
  children: ReactElement | Array<ReactElement>;
}

// TODO TS-migration extend DialogProps, once Dialog is migrated to TS
interface TooltipBaseProps extends VibeComponentProps {
  /**
   * The content displayed inside the tooltip.
   */
  content: ElementContent;
  /**
   * Inline styles applied to the tooltip container.
   */
  style?: CSSProperties;
  /**
   * Class name applied to the tooltip arrow.
   */
  arrowClassName?: string;
  /**
   * Offset values for positioning adjustments.
   */
  moveBy?: MoveBy;
  /**
   * The theme of the tooltip.
   */
  theme?: TooltipTheme;
  /**
   * Function to get the container where the tooltip should be rendered.
   */
  getContainer?: () => HTMLElement;
  /**
   * Delay in milliseconds before hiding the tooltip.
   */
  hideDelay?: number;
  /**
   * Delay in milliseconds before showing the tooltip.
   */
  showDelay?: number;
  /**
   * If true, disables the slide animation of the tooltip.
   */
  disableDialogSlide?: boolean;
  /**
   * The animation type used for showing/hiding the tooltip.
   */
  animationType?: DialogAnimationType;
  /**
   * If true, renders the tooltip without a dialog.
   */
  withoutDialog?: boolean;
  /**
   * The CSS selector of the container where the tooltip should be rendered.
   */
  containerSelector?: string;
  /**
   * Delay in milliseconds before showing the tooltip immediately.
   */
  immediateShowDelay?: number;
  /**
   * If false, hides the arrow of the tooltip.
   */
  tip?: boolean;
  /**
   * If true, the tooltip is shown when the component mounts.
   */
  shouldShowOnMount?: boolean;
  /**
   * If true, hides the tooltip when the reference element is hidden.
   */
  hideWhenReferenceHidden?: boolean;
  /**
   * Callback fired when the tooltip is hidden.
   */
  onTooltipHide?: () => void;
  /**
   * Callback fired when the tooltip is shown.
   */
  onTooltipShow?: () => void;
  /**
   * Custom Popper.js modifiers.
   * https://popper.js.org/docs/v2/modifiers/
   */
  modifiers?: Array<Modifier<unknown>>;
  /**
   * The placement of the tooltip relative to the reference element.
   */
  position?: TooltipPositions;
  /**
   * Events that trigger showing the tooltip.
   */
  showTrigger?: DialogTriggerEvent | Array<DialogTriggerEvent>;
  /**
   * Events that trigger hiding the tooltip.
   */
  hideTrigger?: DialogTriggerEvent | Array<DialogTriggerEvent>;
  /**
   * If true, prevents closing the tooltip when the mouse enters it.
   */
  showOnDialogEnter?: boolean;
  /**
   * Class name applied to the reference wrapper element.
   */
  referenceWrapperClassName?: string;
  /**
   * If true, keyboard focus/blur events behave like mouse enter/leave.
   */
  addKeyboardHideShowTriggersByDefault?: boolean;
  /**
   * If true, controls the open state of the tooltip.
   */
  open?: boolean;
  /**
   * The z-index applied to the tooltip.
   */
  zIndex?: number;
  /**
   * The title of the tooltip.
   */
  title?: string;
  /**
   * The image displayed inside the tooltip.
   */
  image?: string;
  /**
   * The icon displayed next to the title.
   */
  icon?: SubIcon;
  /**
   * The maximum width of the tooltip.
   */
  maxWidth?: number;
  /**
   * The text direction of the tooltip: "ltr", "rtl", or "auto".
   */
  dir?: "ltr" | "rtl" | "auto";
}
// When last tooltip was shown in the last 1.5 second - the next tooltip will be shown immediately
const IMMEDIATE_SHOW_THRESHOLD_MS = 1500;

// Shared state across multiple tooltip instances (i.e last tooltip shown time)
const globalState: { lastTooltipHideTS: number; openTooltipsCount: number } = {
  lastTooltipHideTS: null,
  openTooltipsCount: 0
};

export default class Tooltip extends PureComponent<TooltipProps> {
  wasShown: boolean;
  static positions = TooltipPositionsEnum;
  static hideShowTriggers = HideShowEventEnum;
  static themes = TooltipThemeEnum;
  static animationTypes = AnimationTypeEnum;
  static defaultProps = {
    moveBy: { main: 4, secondary: 0 },
    theme: "dark",
    position: "top",
    hideDelay: 100,
    showDelay: 300,
    disableDialogSlide: true,
    animationType: AnimationTypeEnum.EXPAND,
    withoutDialog: false,
    tip: true,
    hideWhenReferenceHidden: false,
    modifiers: new Array<Modifier<unknown>>(),
    showTrigger: Tooltip.hideShowTriggers.MOUSE_ENTER,
    hideTrigger: Tooltip.hideShowTriggers.MOUSE_LEAVE,
    showOnDialogEnter: true,
    referenceWrapperClassName: "",
    addKeyboardHideShowTriggersByDefault: true,
    open: false
  };
  constructor(props: TooltipProps) {
    super(props);
    this.renderTooltipContent = this.renderTooltipContent.bind(this);
    this.getShowDelay = this.getShowDelay.bind(this);
    this.onTooltipShow = this.onTooltipShow.bind(this);
    this.onTooltipHide = this.onTooltipHide.bind(this);

    this.wasShown = false;
  }

  renderTooltipContent() {
    const { theme, content, className, style, maxWidth, title, image, icon, dir } = this.props;
    if (!content) {
      // don't render empty tooltip
      return null;
    }
    let contentValue;
    if (isFunction(content)) {
      contentValue = content();
    } else if (isValidElement(content)) {
      contentValue = content;
    } else if (typeof content === "string" && content) {
      contentValue = content;
    } else if (Array.isArray(content) && content.length > 0) {
      // allow array of elements
      contentValue = content;
    }

    if (!contentValue) {
      return null;
    }

    return (
      <div
        style={maxWidth ? ({ ...style, "--tooltip-max-width": `${maxWidth}px` } as CSSProperties) : style}
        className={cx(styles.tooltip, getStyle(styles, camelCase(theme)), className)}
        dir={dir}
      >
        {image && <img className={styles.image} src={image} alt="" />}
        <div className={cx(styles.content)}>
          {title && (
            <Flex gap="xs">
              {icon && <Icon iconSize="20" icon={icon} />}
              <div className={styles.title}>{title}</div>
            </Flex>
          )}
          {contentValue}
        </div>
      </div>
    );
  }

  onTooltipShow() {
    if (!this.wasShown) {
      const { onTooltipShow } = this.props;
      globalState.openTooltipsCount++;
      this.wasShown = true;
      onTooltipShow && onTooltipShow();
    }
  }

  onTooltipHide() {
    if (this.wasShown) {
      const { onTooltipHide } = this.props;
      globalState.lastTooltipHideTS = Date.now();
      globalState.openTooltipsCount--;
      this.wasShown = false;
      onTooltipHide && onTooltipHide();
    }
  }

  getTimeSinceLastTooltip() {
    if (globalState.openTooltipsCount > 0) {
      return 0;
    }
    return globalState.lastTooltipHideTS ? Date.now() - globalState.lastTooltipHideTS : Infinity;
  }

  getShowDelay() {
    const { showDelay, immediateShowDelay } = this.props;
    const timeSinceLastTooltip = this.getTimeSinceLastTooltip();
    if ((immediateShowDelay === 0 || immediateShowDelay) && timeSinceLastTooltip < IMMEDIATE_SHOW_THRESHOLD_MS) {
      // showing the tooltip immediately (without animation)
      return {
        showDelay: immediateShowDelay,
        preventAnimation: true
      };
    }
    return {
      showDelay,
      preventAnimation: false
    };
  }

  render() {
    const {
      withoutDialog,
      children,
      forceRenderWithoutChildren,
      theme,
      tip,
      arrowClassName,
      id,
      "data-testid": dataTestId,
      position
    } = this.props;

    if (!children && !forceRenderWithoutChildren) {
      return null;
    }

    if (withoutDialog) {
      return this.renderTooltipContent();
    }

    const content = this.renderTooltipContent;
    const dialogProps = {
      ...this.props,
      position: position,
      "data-testid": dataTestId || getTestId(ComponentDefaultTestId.TOOLTIP, id),
      tooltip: tip,
      content,
      tooltipClassName: cx(styles.arrow, getStyle(styles, theme), arrowClassName),
      onDialogDidHide: this.onTooltipHide,
      onDialogDidShow: this.onTooltipShow,
      getDynamicShowDelay: this.getShowDelay
    };
    return (
      <Dialog {...dialogProps} animationType="expand">
        {children}
      </Dialog>
    );
  }
}
