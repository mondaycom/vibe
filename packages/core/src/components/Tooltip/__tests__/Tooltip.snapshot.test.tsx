import React from "react";
import renderer from "react-test-renderer";
import Tooltip from "../Tooltip";

vi.mock("react-transition-group", () => {
  const FakeTransition = vi.fn(({ children }) => children);
  const FakeSwitchTransition = vi.fn(({ children }) => children);
  const FakeCSSTransition = vi.fn(({ children }) => children);
  return {
    CSSTransition: FakeCSSTransition,
    Transition: FakeTransition,
    SwitchTransition: FakeSwitchTransition
  };
});

vi.useFakeTimers();

describe("Tooltip renders correctly", () => {
  it("with theme", () => {
    const tree = renderer
      .create(
        <Tooltip shouldShowOnMount content="test" theme="primary">
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with position", () => {
    const tree = renderer
      .create(
        <Tooltip shouldShowOnMount content="test" position="left">
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with withoutDialog", () => {
    const tree = renderer
      .create(
        <Tooltip shouldShowOnMount withoutDialog content="test">
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("without arrow", () => {
    const tree = renderer
      .create(
        <Tooltip tip={false} content="test" shouldShowOnMount>
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with hideWhenReferenceHidden", () => {
    const tree = renderer
      .create(
        <Tooltip hideWhenReferenceHidden content="test" shouldShowOnMount>
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with style", () => {
    const tree = renderer
      .create(
        <Tooltip style={{ width: "200px" }} content="test" shouldShowOnMount>
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("with data-testid", () => {
    const tree = renderer
      .create(
        <Tooltip data-testid="test" content="test" shouldShowOnMount>
          <div />
        </Tooltip>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
