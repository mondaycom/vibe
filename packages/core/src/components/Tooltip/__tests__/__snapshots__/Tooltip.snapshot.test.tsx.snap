// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip renders correctly > with data-testid 1`] = `
<body>
  <div>
    <div
      class=""
    />
  </div>
  <span
    class="monday-style-dialog-content-wrapper contentWrapper"
    data-popper-reference-hidden="false"
    data-testid="test"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="contentComponent top hasTooltip expandAppear expandAppearActive"
    >
      <div
        class="tooltip dark"
      >
        <div
          class="content"
        >
          test
        </div>
      </div>
      <div
        class="arrow arrow dark"
        data-placement="top"
        style="position: absolute;"
      />
    </div>
  </span>
</body>
`;

exports[`Tooltip renders correctly > with hideWhenReferenceHidden 1`] = `
<body>
  <div>
    <div
      class=""
    />
  </div>
  <span
    class="monday-style-dialog-content-wrapper contentWrapper"
    data-testid="tooltip"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="contentComponent top hasTooltip expandAppear expandAppearActive"
    >
      <div
        class="tooltip dark"
      >
        <div
          class="content"
        >
          test
        </div>
      </div>
      <div
        class="arrow arrow dark"
        data-placement="top"
        style="position: absolute;"
      />
    </div>
  </span>
</body>
`;

exports[`Tooltip renders correctly > with position 1`] = `
<body>
  <div>
    <div
      class=""
    />
  </div>
  <span
    class="monday-style-dialog-content-wrapper contentWrapper"
    data-popper-reference-hidden="false"
    data-testid="tooltip"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="contentComponent left hasTooltip expandAppear expandAppearActive"
    >
      <div
        class="tooltip dark"
      >
        <div
          class="content"
        >
          test
        </div>
      </div>
      <div
        class="arrow arrow dark"
        data-placement="left"
        style="position: absolute;"
      />
    </div>
  </span>
</body>
`;

exports[`Tooltip renders correctly > with style 1`] = `
<body>
  <div>
    <div
      class=""
    />
  </div>
  <span
    class="monday-style-dialog-content-wrapper contentWrapper"
    data-popper-reference-hidden="false"
    data-testid="tooltip"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="contentComponent top hasTooltip expandAppear expandAppearActive"
    >
      <div
        class="tooltip dark"
        style="width: 200px;"
      >
        <div
          class="content"
        >
          test
        </div>
      </div>
      <div
        class="arrow arrow dark"
        data-placement="top"
        style="position: absolute;"
      />
    </div>
  </span>
</body>
`;

exports[`Tooltip renders correctly > with theme 1`] = `
<body>
  <div>
    <div
      class=""
    />
  </div>
  <span
    class="monday-style-dialog-content-wrapper contentWrapper"
    data-popper-reference-hidden="false"
    data-testid="tooltip"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="contentComponent top hasTooltip expandAppear expandAppearActive"
    >
      <div
        class="tooltip primary"
      >
        <div
          class="content"
        >
          test
        </div>
      </div>
      <div
        class="arrow arrow primary"
        data-placement="top"
        style="position: absolute;"
      />
    </div>
  </span>
</body>
`;

exports[`Tooltip renders correctly > with withoutDialog 1`] = `
<div
  class="tooltip dark"
>
  <div
    class="content"
  >
    test
  </div>
</div>
`;

exports[`Tooltip renders correctly > without arrow 1`] = `
<body>
  <div>
    <div
      class=""
    />
  </div>
  <span
    class="monday-style-dialog-content-wrapper contentWrapper"
    data-popper-reference-hidden="false"
    data-testid="tooltip"
    style="position: absolute; left: 0px; top: 0px;"
  >
    <div
      class="contentComponent top expandAppear expandAppearActive"
    >
      <div
        class="tooltip dark"
      >
        <div
          class="content"
        >
          test
        </div>
      </div>
    </div>
  </span>
</body>
`;
